curve-amm/src/math.rs

```diff
@@ -0,0 +1,163 @@
+    if sum == zero {
+        return Some(zero);
+    }
+    let mut d = sum;
+
+    for _ in 0..255 {
+        let mut d_p = d;
+        for x in xp_f.iter() {
+            // d_p = d_p * d / (x * n)
+            d_p = d_p.checked_mul(d.checked_div(x.checked_mul(n)?)?)?;
+        }
+        let d_prev = d;
+
+        // d = (ann * sum + d_p * n) * d / ((ann - 1) * d + (n + 1) * d_p)
+        let term1 = ann_f.checked_mul(sum)?.checked_add(d_p.checked_mul(n)?)?;
+        let term2 = ann_f
+            .checked_sub(one)?
+            .checked_mul(d)?
+            .checked_add(n.checked_add(one)?.checked_mul(d_p)?)?;
+        d = term1.checked_div(term2)?.checked_mul(d)?;
+        if d > d_prev {
+            let v = d.checked_sub(d_prev)?;
+            if v <= prec {
+                return Some(d);
+            }
+        } else {
+            let v = d_prev.checked_sub(d)?;
+            if v <= prec {
+                return Some(d);
+            }
+        }
+    }
+    None
+}
+/// Here `xp` - coin amounts, `ann` is amplification coefficient multiplied by `n^n`, where
+/// `n` is number of coins.
+///
+/// See https://github.com/equilibrium-eosdt/equilibrium-curve-amm/blob/master/docs/deducing-get_y-formulas.pdf
+/// for detailed explanation about formulas this function uses.
+///
+/// # Notes
+///
+/// Done by solving quadratic equation iteratively.
+///
+/// ```pseudocode
+/// x_1^2 + x_1 * (sum' - (A * n^n - 1) * D / (A * n^n)) = D^(n+1) / (n^2n * prod' * A)
+/// x_1^2 + b * x_1 = c
+///
+/// x_1 = (x_1^2 + c) / (2 * x_1 + b)
+/// ```
+pub fn get_y(i: usize, j: usize, x_f: U256, xp_f: &[U256], ann_f: U256) -> Option<U256> {
+    let zero = U256::zero();
+    let two: U256 = 2_u128.into();
+    let n = (xp_f.len() as u128).into();
+
+    // Same coin
+    if i == j {
+        return None;
+    }
+    // j above n
+    if j >= xp_f.len() {
+        return None;
+    }
+    if i >= xp_f.len() {
+        return None;
+    }
+    let d_f = get_d(xp_f, ann_f)?;
+    let mut c = d_f;
+    let mut s = zero;
+
+    // Calculate s and c
+    // p is implicitly calculated as part of c
+    // note that loop makes n - 1 iterations
+    for (k, xp_k) in xp_f.iter().enumerate() {
+        let x_k: U256;
+        if k == i {
+            x_k = x_f;
+        } else if k != j {
+            x_k = *xp_k;
+        } else {
+            continue;
+        }
+        // s = s + x_k
+        s = s.checked_add(x_k)?;
+        // c = c * d / (x_k * n)
+        c = c.checked_mul(d_f)?.checked_div(x_k.checked_mul(n)?)?;
+    }
+    // c = c * d / (ann * n)
+    // At this step we have d^n in the numerator of c
+    // and n^(n-1) in its denominator.
+    // So we multiplying it by remaining d/n
+    c = c.checked_mul(d_f)?.checked_div(ann_f.checked_mul(n)?)?;
+
+    // b = s + d / ann
+    // We subtract d later
+    let b = s.checked_add(d_f.checked_div(ann_f)?)?;
+    let mut y = d_f;
+
+    for _ in 0..255 {
```

Comment:
Why so magic? Note that this loop iterates from 0 to 254. Is it correct? Anyway, I'd add some meaningful named constant instead of the magic number. And some foundation this value has been chosen based on.

Reply:
Please check original implementation in vyper https://github.com/curvefi/curve-contract/blob/b0bbf77f8f93c9c5f4e415bce9cd71f0cdee960e/contracts/pool-templates/base/SwapTemplateBase.vy#L419
Quote from paper: https://curve.fi/files/stableswap-paper.pdf
```
When a portfolio of coins {xi} is loaded up, we need to calculate D, and we
need to hold this equation true when we perform trades (e.g. swap xi into xj ).
That is done by finding an iterative, converging solution either for D, or for xj
when all other variables are known.
```
Similarly for get_y see https://raw.githubusercontent.com/equilibrium-eosdt/equilibrium-curve-amm/master/docs/deducing-get_y-formulas.pdf
It is basically a quadratic equation and solved with fixed point iteration method. 

I don't know how they have chose 255, but usually it terminates with in 10 iteration. I also think `proptest` for get_y() is good to have confidence in this implementation.